### PR TITLE
Detect system-installed Rt* libraries

### LINKED ISSUE #185: Building against system-installed RtAudio & RtMidi
Some users may prefer to link applications against their system-installed versions of RtAudio & RtMidi, for example to reduce the download size (no submodules required), to avoid having multiple copies of them on their systems - shared library of embedded into applications - or to have a custom/fixed version in use by all of their applications. OPN2BankEditor suffers from the same problem.

Could you adjust your build systems to check for RtAudio & RtMidi? In CMake this should probably involve the use of [`pkg_check_modules` / `pkg_search_module`](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html) for pkg-config support and fall back to regular `find_library`.

### PR DESCRIPTION
**Pull request by OPNA2608**
Closes #185.

- RtAudio & RtMidi: Uses pkg-config in CMake to find these libraries & required settings, falls back to finding the libraries & header directories via CMake-native mechanisms (`find_library` & `find_path`)
- RtAudio: Checks if the found headers have the required function `RtAudio::getCompiledApiName` (introduced with 5.1.0), rejects the detected results if they don't
- JACK & PulseAudio: Fallback if pkg-config is not available
- JACK: Also detect on Darwin

Builds & runs on Ubuntu 18.04 with only RtAudio checked out (RtAudio is still version 5.0.0 there, so only system-RtMidi can be used). I'll test this on Darwin & a more updated Linux some time this week.

Do you want a CMake option to control this detection behaviour? If yes, what should the default value be?

### CODE REVIEW

**Comment by Wohlstand** [comment]
The detection would be fine to be the default, and then have both options:
- fallback into a shipped build of the library
- the manually-set flag that will enforce use the shipped build of the library always

**Comment by Wohlstand** [comment]
P.S. The conflict here is because I had to comment out the code that enables the ASIO support on Windows, that doesn't work properly, unfortunately, and people complained me several problems with it:
- just no audio
- startup crash
So, I desired to disable this at all. At the same time, this is a proprietary technology that is not met by the FOSS community.

**Comment by OPNA2608** [comment]
- Rebased to fix merge conflict
- Added `USE_VENDORED_RTMIDI` and `USE_VENDORED_RTAUDIO` CMake options to disable the library detection (default value is `OFF`)

**Comment by Wohlstand** [comment]
Good! :fox_face: :+1:

File: CMakeLists.txt [hunk 1]
```diff
@@ -271,68 +266,176 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
...
+
+    if(NOT RTAUDIO_FOUND)
+      find_library(RTAUDIO_MANUAL_LIBRARY rtaudio)
+      message("!! RtAudio library: ${RTAUDIO_MANUAL_LIBRARY}")
+      find_path(RTAUDIO_MANUAL_INCLUDE_DIRS NAMES "RtAudio.h" PATH_SUFFIXES "include/rtaudio")
+      message("!! RtAudio header directory: ${RTAUDIO_MANUAL_INCLUDE_DIRS}")
+      set(CMAKE_REQUIRED_INCLUDES ${RTAUDIO_MANUAL_INCLUDE_DIRS})
+      set(CMAKE_REQUIRED_LIBRARIES ${RTAUDIO_MANUAL_LIBRARY})
+      check_cxx_symbol_exists(RtAudio::getCompiledApiName "RtAudio.h" RTAUDIO_MANUAL_SUFFICIENT)
+      if(NOT RTAUDIO_MANUAL_LIBRARY-NOTFOUND AND NOT RTAUDIO_MANUAL_INCLUDE_DIRS STREQUAL RTAUDIO_MANUAL_INCLUDE_DIRS-NOTFOUND AND RTAUDIO_MANUAL_SUFFICIENT)
```
Comment (Wohlstand) [inline CMakeLists.txt:339]: > `NOT RTAUDIO_MANUAL_LIBRARY-NOTFOUND`

Isn't a mistake? :thinking: 
Please put strings into quotes also as they confuse some things here.

Comment (OPNA2608) [reply CMakeLists.txt:339]: I think I misread the docs w/r/t `find_library`'s `<var>-NOTFOUND` :man_facepalming:, will fix.

Comment (Wohlstand) [reply CMakeLists.txt:339]: Usually goes:
```cmake
find_library(KURWA kurwa)

if(KURWA)
     message("Found!")
endif()

if(NOT KURWA)
     message("Kurwa mać!")
endif()
```

Comment (OPNA2608) [reply CMakeLists.txt]: Was quite the hot day when I hacked that together, should be better now. :sweat_smile:

Comment (Wohlstand) [reply CMakeLists.txt]: Now looks much better! :fox_face: :+1:

**Comment by OPNA2608** [comment]
Should I port this to OPN2BankEditor or will you do that?

**Comment by Wohlstand** [comment]
If you want, feel free to try :fox_face: 
Also please apply my patch to disable the ASIO on the OPN2 side too, it still being enabled here.

**Pull request merged**